tests/insert_coins.js

```diff
@@ -0,0 +1,114 @@
+const { ApiPromise, WsProvider, Keyring } = require('@polkadot/api');
+
+const keyring = new Keyring({ type: 'sr25519' });
+
+const fs = require('fs');
+
+// Create a promise API instance of the passed in node address.
+async function createPromiseApi(nodeAddress) {
+    const wsProvider = new WsProvider(nodeAddress);
+    const api = await new ApiPromise({ 
+        provider: wsProvider
+    });
+    await api.isReady;
+    return api;
+}
+const mints = [];
+
+async function main() {
+    let nodeAddress = "ws://127.0.0.1:9801";
+    
+    const api = await createPromiseApi(nodeAddress);
+    let sender = keyring.addFromMnemonic("bottom drive obey lake curtain smoke basket hold race lonely fit walk//Alice");
+
+    const mints_file = '/home/georgi/Desktop/workspace/Manta/precomputed-15k-iterations/precomputed_mints_v3-5';
+    const transfers_file = '/home/georgi/Desktop/workspace/Manta/precomputed-15k-iterations/precomputed_transfers_v3-5';
+    const reclaims_file = '/home/georgi/Desktop/workspace/Manta/precomputed-15k-iterations/precomputed_reclaims_v3-5';
+
+    let mints_offset = 2;
+    let transfers_offset = 4;
+    let reclaims_offset = 4;
+    let total_iterations = 15000;
+
+    let mint_size = 552;
+
+    const mints_buffer = await fs.promises.readFile(mints_file);
+    const transfers_buffer = await fs.promises.readFile(transfers_file);
+    const reclaims_buffer = await fs.promises.readFile(reclaims_file);
+
+    let transactions = [];
+    for(let i = 0; i < 14000; ++i){
+        let mints_start = mints_offset + i * mint_size;
+        const mint = await api.tx.mantaPay.toPrivate(mints_buffer.subarray(mints_start, mint_size + mints_start));
+        transactions.push(mint);
+
+        let transfers_start = transfers_offset + i * (2 * mint_size + transfer_size);
+        const transfer_mint_1 = await api.tx.mantaPay.toPrivate(transfers_buffer.subarray(transfers_start, transfers_start + mint_size));
+        const transfer_mint_2 = await api.tx.mantaPay.toPrivate(transfers_buffer.subarray(transfers_start + mint_size, transfers_start + 2 * mint_size));
+        const transfer = await api.tx.mantaPay.privateTransfer(transfers_buffer.subarray(transfers_start + 2 * mint_size, transfers_start + 2 * mint_size + transfer_size));
+        transactions.push(transfer_mint_1);
+        transactions.push(transfer_mint_2);
+        transactions.push(transfer);
+
+        await api.tx.utility.forceBatch(transactions).signAndSend(sender, {nonce: -1}, ({ events = [], status }) => {
+            if (status.isFinalized) {
+                console.log("tx %i success.", status.nonce);
+            }
+        if (status.isDropped) {
+            console.log("tx %i dropped.", status.nonce);
+            unsub();
+        }
+        if (status.isUsurped) {
+            console.log("tx %i usurped.", status.nonce);
+            unsub();
+        }
+        if (status.isFinalityTimeout) {
+            console.log("tx %i finality timeout.", status.nonce);
```

Comment:
aren't these error cases? Why not abort if this happens?

Reply:
The unsub() terminates, these were error logs.

Reply:
i see this is a weird mix of promise and callback API in this fn, but if that's the way it's used I guess that's fine. I'm mainly confused because i don't see the promise resolving anywhere. so it looks to me that this should wait forever. if that's not the case it's fine

Reply:
i think it's fine to keep these logs, but i'm missing termination of this js with error. if one of them occurs. I guess its just for testing so no need to be too careful, but normally i'd like to detect the sequence failed without looking through 14k lines of logs.

Reply:
maybe at least promote these to log.err()

Reply:
I cleaned it up some more and using log.err() now